File: src/main/java/com/erzbir/mirai/numeron/store/RedisStore.java
package com.erzbir.mirai.numeron.store;

import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;
import redis.clients.jedis.JedisPooled;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Stream;

/**
 * @author Erzbir
 * @Date: 2022/11/30 10:32
 */
public final class RedisStore {
    private static final Object key = new Object();
    private static final JedisPooled client = new JedisPooled("localhost", 6379);
    private static volatile RedisStore INSTANCE;

    private RedisStore() {

    }

    public static RedisStore getInstance() {
        if (INSTANCE == null) {
            synchronized (key) {
                if (INSTANCE == null) {
                    INSTANCE = new RedisStore();
                }
            }
        }
        return INSTANCE;
    }

    public void set(String key, String value, long seconds) {
        if (seconds == -1L) {
            client.set(key, value);
        } else {
            client.setex(key, seconds, value);
        }

    }

    private List<String> toList(Set<String> keys) {
        List<String> list = new ArrayList<>();
        if (keys != null) {
            keys.forEach(v -> list.add(client.get(v)));
        }
        return list;
    }


    public void del(String key) {
        client.del(key);
    }

    public String get(String key) {
        return client.get(key);
    }

    public @NotNull List<String> getPic() {
        return toList(client.keys("pic_*"));
    }

    public @NotNull List<String> getFile() {
        return toList(client.keys("file_*"));
    }

    public @NotNull List<String> getPlain() {
        return toList(client.keys("plain_*"));
    }


    public void removePic(String filename) throws Exception {
        Stream<String> stringStream;
        try {
            stringStream = client.keys("pic_*").stream().filter(v -> Objects.equals(v, filename));
        } catch (Exception e) {
            throw new Exception(e);
        }
        stringStream.forEach(client::del);
    }

    public void removeFile(@NotNull String filename) throws Exception {
        Stream<String> stringStream;
        try {
            stringStream = client.keys("file_*").stream().filter(v -> Objects.equals(v, filename));
        } catch (Exception e) {
            throw new Exception(e);
        }
        stringStream.forEach(client::del);
    }

    public void removePlain(String filename) throws Exception {
        Stream<String> stringStream;
        try {
            stringStream = client.keys("plain_*").stream().filter(v -> Objects.equals(v, filename));
        } catch (Exception e) {
            throw new Exception(e);
        }
        stringStream.forEach(client::del);

    }

    public void useClient(Function1<JedisPooled, Object> function1) {
        function1.invoke(client);
    }
}

File: src/main/java/com/erzbir/mirai/numeron/store/RedisStore.kt
package com.erzbir.mirai.numeron.store

import redis.clients.jedis.Jedis

object RedisStore {
    private val client = Jedis("localhost", 6379)

    fun set(key: String, value: String, seconds: Long = -1L) {
        if (seconds == -1L) {
            client.set(key, value)
        } else {
            client.setex(key, seconds, value)
        }
    }

    fun del(key: String) {
        println("del:$key")
        client.del(key)
    }

    fun get(key: String): String? {
        return client.get(key)
    }

    fun getPic(): List<String> {
        return client.keys("pic_*").map {
            client.get(it)
        }
    }

    fun getFile(): List<String> {
        return client.keys("file_*").map {
            client.get(it)
        }
    }

    fun getPlain(): List<String> {
        return client.keys("plain_*").map {
            client.get(it)
        }
    }

    fun removePic(filename: String) {
        client.keys("pic_*").filter {
            client.get(it) == filename
        }.forEach {
            client.del(it)
        }
    }

    fun removeFile(filename: String) {
        client.keys("file_*").filter {
            client.get(it) == filename
        }.forEach {
            client.del(it)
        }
    }

    fun removePlain(filename: String) {
        client.keys("plain_*").filter {
            client.get(it) == filename
        }.forEach {
            client.del(it)
        }
    }

    fun useClient(r: (client: Jedis) -> Unit) {
        r(client)
    }
}